atomic_reactor/utils/resource_manager.py

```diff
@@ -67,186 +166,378 @@ def username(self):
+        """ Get slots on host which are in free state """
+        logger.debug("%s: retrieve list of available slots", self.hostname)
+        available_slots = []
+        for slot_id in range(self.slots):
+            if not self.is_free(slot_id):
+                logger.debug("%s: slot %s is not free", self.hostname, slot_id)
+                continue
+            available_slots.append(slot_id)
 
         return available_slots
 
-    def get_cmd_to_read_from_json(self, file_name: str):
-        """ Create a command to retrieve content of json file from remote host """
-        return ['cat', os.path.join(RELATIVE_PATH, file_name)]
-
-    def get_cmd_to_write_empty_slot(self, slot_id: int):
-        """ Create a command to unlock slot on remote host """
-        return ['echo', '{}', '>', os.path.join(RELATIVE_PATH, f'slot_{slot_id}.json')]
 
-    def get_cmd_to_write_slot_in_progress(self, file_name: str, payload: str):
-        """ Create a command to lock slot for pipelinerun on remote host """
-        filepath = os.path.join(RELATIVE_PATH, file_name)
+class HostSlot:
 
-        # TBD: Test with remote host that this is actually working with Paramiko
-        _cmd = (f'if grep -qw {{}} {filepath} ; then echo {quote(payload)} > '
-                f'{filepath} ;else exit 1; fi')
-        return f'flock {filepath} -c {quote(_cmd)}'
+    def __init__(self, host: RemoteHost, session: SSHRetrySession, slot_id: int):
+        """ Instantiate host slot with remote host instance, an ssh session and slot id
 
-    @backoff.on_exception(
-        backoff.expo,
-        LockError,
-        factor=BACKOFF_FACTOR,
-        max_tries=MAX_RETRIES,
-        jitter=None,  # use deterministic backoff, do not apply random jitter
-        logger=logger,
-    )
-    def lock(self, slot_id: int, prid: str):
-        logger.debug("Locking slot %s on host %s with pipelinerun ID %s",
-                     slot_id, self.hostname, prid)
-
-        payload = json.dumps({'prid': prid, 'locked': str(datetime.utcnow().isoformat())})
-
-        with self._ssh_session() as client:
-            cmd = self.get_cmd_to_write_slot_in_progress(f'slot_{slot_id}.json', payload)
-            _, _, exitcode = self._ssh_run_remote_cmd(client, cmd)
+        :param host: RemoteHost, RemoteHost instance
+        :param session: SSHRetrySession, SSHRetrySession instance
+        :param slot_id: int, slot ID
+        """
+        self.host = host
+        self.hostname = host.hostname
+        self.session = session
+        self.id = slot_id
+        self.path = os.path.join(self.host.slots_dir, f"slot_{slot_id}")
 
-            if exitcode != 0:
-                raise LockError(f'Attempt to lock slot {slot_id} with pipelinerun ID {prid} '
-                                'returned non-zero exitcode')
+    @property
+    def _data(self) -> SlotData:
+        content = self._read()
+        return SlotData.from_string(content)
```

Comment:
Yes, it impacts the performance, currently I think it's acceptable, this will make it easy implement other APIs without worrying about inconsistent data, I'll do some testing and thinking on how can do this better, but probably not in this PR.

Reply:
Ok, should be fine for now